xactserver/src/lib.rs

```diff
@@ -33,3 +34,60 @@ pub enum XsMessage {
     Prepare(proto::PrepareRequest),
     Vote(proto::VoteRequest),
 }
+
+#[derive(Clone, Debug, PartialEq)]
+pub enum XactStatus {
+    Unexecuted,
+    Waiting,
+    Committed,
+    Rollbacked(RollbackInfo),
+}
+
+impl XactStatus {
+    pub fn is_terminal(&self) -> bool {
+        matches!(self, XactStatus::Committed | XactStatus::Rollbacked(_))
+    }
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct RollbackInfo(NodeId, RollbackReason);
+
+#[derive(Clone, Debug, PartialEq)]
+pub enum RollbackReason {
+    Db(DbError),
+    Other(String),
```

Comment:
The Other attribute might be extra, we could simply add the information in the "detail" part and that way make sure everything is a part of the DbError object?

Reply:
I mean we could have a fixed code for an unknown error and add the string into detail?

Reply:
The `DbError` stores errors returned from postgresql, all fields in this errors (errorcode, message, details, ect.) are filled by postgres, so we need a separate type of error for those that are external to postgres (connection failure, errors in the `tokio-postgres` library, errors in the xactserver, etc.).